New | Connectors Documentation (#58)
mia-platform/documentation: docs/fast_data/connectors/debezium_cdc.md

---
id: debezium_cdc
title: Debezium Server CDC
sidebar_label: Debezium Server CDC
---

> In this page is describe how to configure Debezium Server to connect with all the supported databases.

Reviewer: ```suggestion
This page provides instructions on how to configure the Debezium Server to connect with all the databases that are supported.
```

## Overview

As previously introduced, in the Mia-Platform Marketplace are available a set of plugins that repackages Debezium Server image, adding the JMX Prometheus Exporter to expose collected metrics regarding the connector.

> These plugins are shipped with a predefined configuration tailored for the database of interest to speed up the configuration and deploy phase. This configuration, saved in a file named `application.properties` is mounted as a config map on the path `/debezium/conf` within the container. This leverage the mechanism [Quarkus](https://quarkus.io/) framework, on which Debezium Server is based on, adopts to extend the base configuration of a service.

Reviewer: ```suggestion
These plugins are shipped with a predefined configuration tailored for the database of interest to speed up the configuration and deployment phase. This configuration, saved in a file named `application.properties` is mounted as a config map on the path `/debezium/conf` within the container. This leverages the mechanism of [Quarkus](https://quarkus.io/) framework, on which Debezium Server is based, in order to extend the base configuration of a service.
```

:::info

> Current Debezium Server plugin requires a dependency on [Redis](https://redis.io/) for storing the metadata regarding the offset and schema management. More details can be found below or in the official Documentation. 

Reviewer: can we add a link to the official documentation? and also add a link to "can be found below" with the paragraph linked

Author: I rephrased this part, since actually there is no clear explanation on the official documentation explaining offsets metadata management.

Reviewer: ok, I would put into the same :::note paragraph instead of having the info box and just after the note, since they are related

| DATABASES_LIST            | true     | Comma separated list of database names or regular expression that defines which databases should be monitored                | -             |  
| TABLES_LIST               | true     | Comma separated list of tables names (with their schema) or regular expression that defines which tables should be monitored | -             |  
| KAFKA_BROKERS             | true     | Comma separated list of nodes address belonging to a Kafka cluster                                                           | -             |
| KAFKA_USERNAME            | true     | The Kafka username                                                                                                           | -             |
| KAFKA_PASSWORD            | true     | The Kafka password                                                                                                           | -             |
| KAFKA_CLIENT_ID           | false    | Client identifier employed by this application                                                                               | -             |

> In the following section are provided a configuration example for each supported database and some instructions on how to configure and enable the database itself for CDC operations. Furthermore, it is clarified why Debezium Server CDC currently requires to be connected also to an instance of Redis.
> 
> ## Debezium Server Configuration
> 
> In these paragraphs are detailed the configurations that have tested out with the Debezium Server plugin.  
> Currently, the default sink configuration is prefilled with Kafka properties, since it is the main event streaming platform supported by Fast-Data. Nonetheless, it is possible to fill it with any of the Debezium supported sink configuration as explained in the [official documentation](https://debezium.io/documentation/reference/2.1/operations/debezium-server.html#_sink_configuration). 

Reviewer: ```suggestion
## Debezium Server Configuration

In this section, some configuration examples for each supported database and some instructions on how to configure and enable the database itself for CDC operations are provided. Furthermore, it is clarified why Debezium Server CDC currently requires to be connected also to an instance of Redis.

In the next paragraphs, the configurations that have been tested out with the Debezium Server plugin are detailed.  
Currently, the default sink configuration is prefilled with Kafka properties, since it is the main event streaming platform supported by Fast-Data. Nonetheless, it is possible to fill it with any of the Debezium-supported sink configurations as explained in the [official documentation](https://debezium.io/documentation/reference/2.1/operations/debezium-server.html#_sink_configuration). 
```

### MySQL

CDC on MySQL can exploit the database binlog ([Binary Log](https://dev.mysql.com/doc/refman/8.0/en/binary-log.html)), which keep tracks of all the changes occurring within the database. In order to allow Debezium execute its replication operations it is therefore necessary to enable the binlog and create a user which can read the database of interest.

Regarding enabling MySQL binary log, we recommend checking out Debezium [instructions](https://debezium.io/documentation/reference/2.1/connectors/mysql.html#enable-mysql-binlog) and refer to the [MySQL manual](https://dev.mysql.com/doc/refman/8.0/en/replication-options-binary-log.html) since it is necessary to edit the server configuration files. 

#### Database User and Permission Configuration

> Below are provided instead the commands to create a user dedicated for CDC operations, granting the proper permissions.

Reviewer: ```suggestion
The commands to create a dedicated user to CDC operations, granting the proper permissions, are provided here below.
```

```sql
-- CDC Configurations (replace variables identified with double curly braces with their correct value)
CREATE USER '{{DEBEZIUM_DB_USERNAME}}' IDENTIFIED WITH mysql_native_password BY '{{DEBEZIUM_DB_PASSWORD}}';

-- these are global privileges that must be granted on all the databases to be monitored
GRANT SELECT,RELOAD,SHOW DATABASES,REPLICATION SLAVE,REPLICATION CLIENT ON *.* TO '{{DEBEZIUM_DB_USERNAME}}';

-- ensure privileges are set
FLUSH PRIVILEGES;
```

#### Debezium Service Configuration

> Once the database is configured, it is possible to configure the Debezium Server instance, filling in the _source_ and _sink_ configuration. Below it is provided an example using MySQL as _source_, Redis as offsets and metadata store and Kafka as _sink_.  

Reviewer: ```suggestion
Once the database is configured, it is possible to configure the Debezium Server instance, filling in the _source_ and _sink_ configuration. An example using MySQL as _source_, Redis as offsets and metadata store and Kafka as _sink_ is provided here below.
```

debezium.sink.kafka.producer.key.serializer = org.apache.kafka.common.serialization.StringSerializer
debezium.sink.kafka.producer.value.serializer = org.apache.kafka.common.serialization.StringSerializer

debezium.sink.kafka.producer.acks = all
debezium.sink.kafka.producer.max.in.flight.requests.per.connection = 5
debezium.sink.kafka.producer.enable.idempotence = true
debezium.sink.kafka.producer.min.insync.replicas = 2
debezium.sink.kafka.producer.retries = 2147483647
debezium.sink.kafka.producer.compression.type = snappy

# Format configuration
debezium.format.key = json
debezium.format.value = json
```

</details>

:::note
For an in depth explanation of the particular database configuration, we suggest reading the [official documentation](https://debezium.io/documentation/reference/2.1/connectors/mysql.html).
:::

### Oracle DB

Debezium CDC over Oracle Database exploits the [LogMiner](https://docs.oracle.com/en/database/oracle/oracle-database/21/sutil/oracle-logminer-utility.html) Oracle component to read and analyze the _redo_ logs (either the `archive` or the `online` logs). In order to allow Debezium carrying out its task the `ARCHIVELOG` feature has to be activated.

> Below is reported the instructions and references to instruct an [AWS RDS for Oracle](https://aws.amazon.com/rds/oracle/) instance to be ready for CDC operations.

Reviewer: ```suggestion
The instructions and references to instruct an [AWS RDS for Oracle](https://aws.amazon.com/rds/oracle/) instance to be ready for CDC operations are reported here below.
```

------------
ARCHIVELOG
```

In case `LOG_MODE` is not set to `ARCHIVELOG` it means the feature is not enabled. Consequently, it has to be activated executing the following AWS RDS administration procedures:

```sql
# define log files retention
exec rdsadmin.rdsadmin_util.set_configuration('archivelog retention hours', 24);

# enable further logs to be written for supporting LogMiner operations
exec rdsadmin.rdsadmin_util.alter_supplemental_logging('ADD');
```

:::caution
In order to actually activate `ARCHIVELOG` feature on AWS RDS for Oracle, the database backup must be enabled (it may require a database restart). This instructs the system to retain also the redo file for the configured amount of time, so that the CDC features work as expected.

Current, one day backup retention does not introduce any new cost on the AWS plan, as explained on [AWS blog](https://aws.amazon.com/blogs/database/demystifying-amazon-rds-backup-storage-costs/). 
:::

:::note
For further details, please check out [Debezium for Oracle documentation](https://debezium.io/documentation/reference/2.1/connectors/oracle.html#_preparing_the_database).
:::

#### Database User and Permission Configuration

> Here are provided the commands that create a database user dedicated to Debezium Server and that grants the appropriate permissions.

Reviewer: ```suggestion
In this paragraph, the commands that create a database user dedicated to Debezium Server and that grant the appropriate permissions are listed.
```

> Redo logs may be sized too small using the default mining strategy, consider increasing redo log sizes to a minimum of 500MB.

Following this recommendation, the size of Redo log files should be increased, for example to 512MB each. The steps to achieve this are:
- create a number of new log files, corresponding to the one initially available, each of them with the new required size
- rotate the active log file until one of the newer files is selected
- deactivate the previous log files so that they can be deleted without issues
- drop the default log files

These operations are clearly described on the [AWS RDS Documentation](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Appendix.Oracle.CommonDBATasks.Log.html#Appendix.Oracle.CommonDBATasks.ResizingRedoLogs), so that we advise to read such guide to perform these actions on the database.

:::info
It is possible to find a reference for each of AWS RDS Administration tasks described above in the AWS RDS [Administration documentation](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Appendix.Oracle.CommonDBATasks.html). 
:::

#### Debezium Service Configuration

> Once the database is configured, it is possible to configure the Debezium Server instance, filling in the _source_ and _sink_ configuration. Below it is provided an example using Oracle as _source_, Redis as offsets and metadata store and Kafka as _sink_.

Reviewer: ```suggestion
Once the database is configured, it is possible to configure the Debezium Server instance, filling in the _source_ and _sink_ configuration. An example using Oracle as _source_, Redis as offsets and metadata store and Kafka as _sink_ is provided here below.
```

## Overview

> As previously introduced, in the Mia-Platform Marketplace are available a set of plugins that repackages Debezium Server image, adding the [JMX Prometheus Exporter](https://github.com/prometheus/jmx_exporter) to expose collected metrics regarding the connector.

Reviewer 2: Should we set the nofollow for crawlers on these links? @akelity

Author: Should we open an issue in this repo, so that we can keep track of the SEO outside this PR?

Reviewer 3: Sorry for the delay in the response, it's something we have to manage, let's open the issue and proceed from there.
